service/qos/service_level_controller.cc

```diff
@@ -687,6 +688,89 @@ future<> service_level_controller::unregister_subscriber(qos_configuration_chang
     return _subscribers.remove(subscriber);
 }
 
+static sstring describe_service_level(std::string_view sl_name, const service_level_options& sl_opts) {
+    using slo = service_level_options;
+
+    utils::small_vector<sstring, 2> opts{};
+    
+    const sstring sl_name_formatted = cql3::util::maybe_quote(sl_name);
+
+    if (auto maybe_timeout = std::get_if<lowres_clock::duration>(&sl_opts.timeout)) {
+        // According to the documentation, `TIMEOUT` has to be expressed in milliseconds
+        // or seconds. It is therefore safe to use milliseconds here.
+        const auto timeout = std::chrono::duration_cast<std::chrono::milliseconds>(*maybe_timeout);
+        opts.push_back(seastar::format("TIMEOUT = {}", timeout));
+    }
+
+    switch (sl_opts.workload) {
+        case slo::workload_type::batch:
+            opts.push_back("WORKLOAD_TYPE = 'batch'");
+            break;
+        case slo::workload_type::interactive:
+            opts.push_back("WORKLOAD_TYPE = 'interactive'");
+            break;
+        case slo::workload_type::unspecified:
+            break;
+        case slo::workload_type::delete_marker:
+            // `slo::workload_typ::delete_marker` is only set temporarily. When a service level
+            // is actually created, it never has this workload type set anymore.
+            on_internal_error(sl_logger, "Unexpected workload type");
+    }
+
+    if (opts.size() == 0) {
+        return seastar::format("CREATE SERVICE LEVEL {};", sl_name_formatted);
+    }
+
+    return seastar::format("CREATE SERVICE LEVEL {} WITH {};", sl_name_formatted, fmt::join(opts, " AND "));
+}
+
+future<std::vector<cql3::description>> service_level_controller::describe_service_levels() const {
+    std::vector<cql3::description> result{};
+
+    if (is_v2()) {
```

Comment:
Thank you for taking a look! What's the demerit of the branch? Won't getting rid of it potentially introduce a consistency issue if e.g. we create a service level and attach it to a role before the cache gets updated?

Reply:
> Thank you for taking a look! What's the demerit of the branch? 

Currently the else branch is doing full scan internal query instead of just accessing a map.

> Won't getting rid of it potentially introduce a consistency issue if e.g. we create a service level and attach it to a role before the cache gets updated?

I would assume the user need to make sure the cluster is in consistent state as part of pre-backup procedure.

Cc @piodul

Reply:
I don't think it's worth to have this branch. It can only happen in a cluster which is either in the middle of rolling upgrade or the migration to topology on raft was not started. I don't think it is a good idea to do backup in such state.

Without schema, auth and service levels on raft you cannot guarantee that you will have consistent results - all this data is put into separate tables and you don't really have a way to atomically fetch them all at once. So, I think it is fine if we read from the cache that is updated every 10s.

Ensuring that this branch is covered by a test would also be a bit painful - you would need to have a test which specifically sets up this kind of scenario just to cover this unlikely branch.